docs: cluster + replica explanation (#12650)
MaterializeInc/materialize: doc/user/content/sql/create-cluster-replica.md

---
title: "CREATE CLUSTER REPLICA"
description: "`CREATE CLUSTER REPLICA` provisions physical resources to perform computations."
menu:
  main:
    parent: commands
---

`CREATE CLUSTER REPLICA` provisions physical resources to perform computations.

## Conceptual framework

> Where [cluster replicas](/overview/key-concepts#clusters) represent the logical

Reviewer: ```suggestion
Where [clusters](/overview/key-concepts#clusters) represent the logical
```

set of dataflows you want to maintain, cluster replicas are their physical
counterparts. Cluster replicas are where Materialize actually creates and
maintains dataflows.

Each cluster replica is essentially a clone, constructing the same dataflows.
Each cluster replica receives a copy of all data that comes in from sources its

> dataflows use, and uses the data to performs identical computations. This design

Reviewer: ```suggestion
dataflows use, and uses the data to perform identical computations. This design
```

provides Materialize with active replication, and so long as one replica is
still reachable, the cluster continues making progress.

This also means that all of a cluster's dataflows contend for the same resources
on each replica. This might mean, for instance, that instead of placing many
complex materialized views on the same cluster, you choose some other
distribution, or you replace all replicas in a cluster with more powerful
machines.

## Syntax

{{< diagram "create-cluster-replica.svg" >}}

### `replica_option`

{{< diagram "replica-option.svg" >}}

Field | Use
------|-----
_cluster_name_ | The cluster whose resources you want to create an additional computation of.
_replica_name_ | A name for this replica.
_replica_option_ | This replica's specified [options](#replica_option).

> _size_ | A "size" (i.e. number of timely workers) for a managed replica––valid values are powers of 2.

Reviewer 2: I'm not sure we want to commit that the sizes are number of Timely works and powers of two. For our cloud offering, we certainly wouldn't use the builtin default values but rather pass a custom `--cluster-replica-sizes` parameter. In the past, we had the idea that the sizes would be something like t-shirt sizes, but let's sync with Cloud to get an agreement on this.

Reviewer 3: Cloud already has a take! The sizes are `xsmall`, `small`, `medium`, `large`, `xlarge`, `2xlarge`, `3xlarge`, `4xlarge`, `5xlarge`, and `6xlarge`.

https://github.com/MaterializeInc/cloud/blob/4077c0365c33798f993940589e4843a0d1920c99/Pulumi.production.yaml#L93-L102

_az_ | If you want the replica to reside in a specific availability zone. You must specify an AWS [availability zone] in either `us-east-1` or `eu-west-1`, e.g. `use1-az1`.
_address_ | The address of the remote machine you want to run this replica.

## Details

### Deployment options

Materialize is an active-replication-based system, which means you expect each
cluster replica to have the same working set.

With this in mind, when building your Materialize deployment, you can change its
performance characteristics by...

Action | Outcome
---------|---------

> Increase all replicas' memory + CPU | Ability to maintain more dataflows or more complex dataflows

Reviewer: ```suggestion
Increase all replicas' size | Ability to maintain more dataflows or more complex dataflows
```

Reviewer: Reflecting Nikhil's suggestion.

Add replicas to a cluster | Greater tolerance to replica failure

### Homogeneous vs. heterogeneous hardware provisioning

Because Materialize uses active replication, all replicas will be asked to do
the same work, irrespective of their resources.

For the most stable performance, we recommend provisioning the same class of
hardware for all replicas.

However, it is possible to provision multiple type of hardware in the same
cluster. In these cases, the slower machines will likely be continually burdened
with a backlog of work. If all of the faster machines become unreachable, the
system might experience delays in replying to requests while the slower machines
catch up to the last known time that the faster machines had computed.

## Example

```sql

> CREATE CLUSTER REPLICA c1.r1 SIZE = '1';

Reviewer 2: Re. cluster sizes.
